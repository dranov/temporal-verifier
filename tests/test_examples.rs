// Copyright 2022 VMware, Inc.
// SPDX-License-Identifier: BSD-2-Clause

use std::{
    ffi::OsStr,
    path::{Path, PathBuf},
    process::Command,
};

use walkdir::WalkDir;

fn get_tests(dir: &str) -> Vec<PathBuf> {
    WalkDir::new(format!("examples/{dir}"))
        .sort_by_file_name()
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|entry| {
            entry.file_type().is_file() && entry.path().extension() == Some(OsStr::new("fly"))
        })
        .map(|entry| entry.path().to_path_buf())
        .collect()
}

fn path_test_name(path: &Path) -> String {
    path.strip_prefix("examples")
        .unwrap()
        .to_string_lossy()
        .to_string()
}

#[test]
fn success() {
    for path in get_tests("success") {
        let out = Command::new("./target/debug/temporal-verifier")
            .arg("--color=never")
            .arg(path.as_os_str())
            .output()
            .expect("could not run verifier");
        assert!(
            out.status.success(),
            "verification failed for {}",
            path.display()
        );
    }
}

#[test]
fn failing_snapshot() {
    for path in get_tests("fail") {
        println!("{}", path.display());
        let out = Command::new("./target/debug/temporal-verifier")
            .arg("--color=never")
            .arg(path.as_os_str())
            .output()
            .expect("could not run verifier");
        let stderr = String::from_utf8(out.stderr).expect("non-utf8 output");
        insta::assert_display_snapshot!(path_test_name(&path), stderr);
    }
}
