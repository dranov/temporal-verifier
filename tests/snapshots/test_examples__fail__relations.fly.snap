---
source: tests/test_examples.rs
expression: stderr
---
verification errors:
error: assertion failure
   ┌─ examples/fail/relations.fly:13:1
   │
13 │ assert forall a:A b:A. q(a, b) -> a = b
   │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   │
   = counter example:
     universe A: A!val!0 A!val!1 A!val!2 A!val!3
     a0 = A!val!0
     p = (lambda (x!0) (and (= x!0 A!val!0) (not (= x!0 A!val!2)) (not (= x!0 A!val!3))))
     q = (lambda (x!0 x!1) (or (and (= x!0 A!val!2) (not (= x!0 A!val!3)) (= x!1 A!val!2) (not (= x!1 A!val!3))) (and (not (= x!0 A!val!0)) (not (= x!0 A!val!2)) (not (= x!0 A!val!3)) (not (= x!1 A!val!0)) (not (= x!1 A!val!2)) (not (= x!1 A!val!3))) (and (= x!0 A!val!0) (not (= x!0 A!val!2)) (not (= x!0 A!val!3)) (= x!1 A!val!3)) (and (= x!0 A!val!0) (not (= x!0 A!val!2)) (not (= x!0 A!val!3)) (= x!1 A!val!0) (not (= x!1 A!val!2)) (not (= x!1 A!val!3))) (and (not (= x!0 A!val!0)) (not (= x!0 A!val!2)) (not (= x!0 A!val!3)) (= x!1 A!val!2) (not (= x!1 A!val!3))) (and (= x!0 A!val!3) (= x!1 A!val!3))))


